Create a special optimizer pipeline for constant INSERTs

ggevay opened the pull request:
This PR fixes this regression in INSERT performance: https://github.com/MaterializeInc/database-issues/issues/8801
It adds a new, very simple optimizer pipeline, whose job is just to take care of constant INSERTs.

The test that regressed in the issue by 10-20% compared to v0.125.3 is now 30-40% faster compared to v0.125.3:
https://buildkite.com/materialize/nightly/builds/10584

### Motivation

  * This PR fixes a recognized performance regression: https://github.com/MaterializeInc/database-issues/issues/8801

### Tips for reviewer

The first commit is just some trivial cleanup, can be reviewed separately.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

def- commented:
> The test that regressed in the issue by 10-20% compared to v0.125.3 is now 30-40% faster compared to v0.125.3:

Wonderful, thank you for checking. I'll verify locally with a larger scale too.

def- commented:
Fully nightly run triggered, good for me if green since this should be used across a lot of existing tests: https://buildkite.com/materialize/nightly/builds/10588

ggevay commented:
Well, I've also run the test locally, and it showed a 15-16% regression. What could be the reason for it behaving differently locally? Also, I'm curious to see what your local runs show.

I did
```
bin/mzcompose --find feature-benchmark down && bin/mzcompose --find feature-benchmark run default --scenario InsertMultiRow --other-tag v0.125.3 --scale=5
```
as recommended in the issue.

def- commented:
The main difference is that in Nightly this scenario runs with scale 4, while this is running it with scale 5 (10 times larger).

frankmcsherry commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -2647,7 +2647,7 @@ impl Coordinator {
             let optimizer_config = optimize::OptimizerConfig::from(self.catalog().system_config());
 
-            // Build an optimizer for this VIEW.
+            // (`optimize::view::Optimizer` has a special case for INSERTs.)
```
I think the more germane comment is that it has a special case for *constants*. At least, I hope this to be true as part of merging the PR! :D

ggevay replied:
Ok, changed it.

frankmcsherry commented on src/adapter/src/optimize.rs:
```diff
@@ -359,6 +359,22 @@ fn optimize_mir_local(
     Ok::<_, OptimizerError>(expr)
 }
 
+/// This calls a simple optimizer that is able to fully optimize constant INSERTs. For more details,
+/// see [mz_transform::Optimizer::constant_insert_optimizer].
+#[mz_ore::instrument(target = "optimizer", level = "debug", name = "constant_insert")]
+fn optimize_mir_constant_insert(
```
Why only `INSERT` statements, rather than any constant view? I think we'll want to remove the `_insert`, as it is called on all expressions, not just inserts. Which is great, btw, just .. confusing to call out inserts as what's going on here.

frankmcsherry replied:
Also, what does it mean to "fully optimize" constant expressions? It might be a moment to spell it out as clearly as possible and make sure we do that. E.g. "any expression all of whose leaves are constants, of size small enough to be inlined and folded" or something.

ggevay replied:
> Why only INSERT statements, rather than any constant view? I think we'll want to remove the _insert, as it is called on all expressions, not just inserts. Which is great, btw, just .. confusing to call out inserts as what's going on here.

Ok, made it more general: it now calls `fold_constants_fixpoint()`, which generally works for constants.

> Also, what does it mean to "fully optimize" constant expressions? It might be a moment to spell it out as clearly as possible and make sure we do that. E.g. "any expression all of whose leaves are constants, of size small enough to be inlined and folded" or something.

I've spelled this out on `fold_constants_fixpoint`, and made some other comments point to that.

frankmcsherry commented on src/adapter/src/optimize/view.rs:
```diff
@@ -7,7 +7,8 @@
 // by the Apache License, Version 2.0.
 
-//! Optimizer implementation for `CREATE VIEW` statements.
+//! Optimizer implementation for `CREATE VIEW` statements and other misc statements, such as
+//! `INSERT`.
```
I'm not sure the additional context helps much. Sure we can just say "for relational expressions" or something? We're still in a file called `view.rs`, so .. there's more to do if we really want to clean things up, but I think "and other misc statements" could be tightened up or removed.

ggevay replied:
Done

frankmcsherry replied:
Looks the same to me at the moment. Not the most important detail, but also perhaps something wasn't pushed / refreshed?

ggevay replied:
Oops, sorry. Changed now to
```
//! An Optimizer that
//! 1. Optimistically calls `optimize_mir_constant`.
//! 2. Then, if we haven't arrived at a constant, it calls `optimize_mir_local`, i.e., the
//!    logical optimizer.
//!
//! This is used for `CREATE VIEW` statements and in various other situations where no physical
//! optimization is needed, such as for `INSERT` statements.
```

frankmcsherry commented on src/adapter/src/optimize/view.rs:
```diff
@@ -52,13 +56,24 @@ impl Optimize<HirRelationExpr> for Optimizer {
             TransformCtx::local(&self.config.features, &self.typecheck_ctx, &mut df_meta);
-        let expr = optimize_mir_local(expr, &mut transform_ctx)?;
+
+        // This optimizer is also used for INSERTs (in addition to VIEWs). Therefore, we first call
+        // a very simple optimizer, which should fully take care of constant INSERTs.
+        expr = optimize_mir_constant_insert(expr, &mut transform_ctx)?;
```
I recommend to remove the reference to inserts, and just have `optimize_mir_constants` which does a general thing for all expressions. The comment can call out why it is important (performance on constant inserts).

ggevay replied:
Done

frankmcsherry commented on src/transform/src/lib.rs:
```diff
@@ -819,6 +819,40 @@ impl Optimizer {
         }
     }
 
+    /// Builds a tiny optimizer, which is only suitable for optimizing the MIR plans that come from
+    /// constant INSERTs, e.g.
```
Again, I don't know why the INSERT is here. All constant expressions, surely?

frankmcsherry commented on src/transform/src/lib.rs:
```diff
@@ -819,6 +819,40 @@ impl Optimizer {
+    pub fn constant_insert_optimizer(_ctx: &mut TransformCtx) -> Self {
+        let transforms: Vec<Box<dyn Transform>> = vec![
+            Box::new(NormalizeLets::new(false)),
+            Box::new(canonicalization::ReduceScalars),
```
I don't follow why `ReduceScalars` is here. It doesn't do relation constant folding, but the other two do.

ggevay replied:
Just in case a user writes something like
```
insert into t values (1+1+1+1);
```

ggevay replied:
(This is now included in `fold_constants_fixpoint()`.)

frankmcsherry commented on src/transform/src/lib.rs:
```diff
@@ -819,6 +819,40 @@ impl Optimizer {
+            }),
+        ];
+        Self {
+            name: "constant_insert_optimizer",
```
`"constast_optimizer"`

ggevay replied:
Done

frankmcsherry reviewed (changes requested):
The code seems fine, but the structure, naming, and comments have a strong binding to `INSERT` statements, but apply the logic beyond insert statements. It seems to generally treat constant expressions better, independent of whether they are inserts or other statements. I think we should make sure the code reflects that, vs treating it as a special case.

ggevay commented:
Ok, figured out that it was still showing a regression locally for me because I was running it with a larger scale (as mentioned by Dennis), and the query was so large that `FoldConstants` gave up due to hitting `FOLD_CONSTANTS_LIMIT`, so then I was still running the full logical optimizer. I think there is no need to put more effort into optimizing such large queries, so I'll just leave that as is.

ggevay commented:
Thanks for the review @frankmcsherry! I've addressed the comments.

I'm running Nightly again: https://buildkite.com/materialize/nightly/builds/10589
The result I'm expecting is that the speedup will be slightly less now that we are calling `fold_constants_fixpoint()` instead of a custom sequence of transforms, but I think we'll still have a speedup.

ggevay commented:
Locally, `scale=4` is showing me 10-20% _faster_.

frankmcsherry commented:
> So this fixes the performance for smaller inserts (1000 rows) but not with many rows (10000 rows), interesting.

@antiguru observed that with large constants, much of the time of `FoldConstants` is in trying to determine if there is a column that forms a unique key.

frankmcsherry commented on src/transform/src/lib.rs:
```diff
@@ -565,11 +567,11 @@ pub fn fold_constants_fixpoint() -> Fixpoint {
         name: "fold_constants_fixpoint",
         limit: 100,
         transforms: vec![
+            Box::new(NormalizeLets::new(false)),
```
I don't understand why this moved. We end up with a thing that may not be normalized, but I don't see why we want that.

ggevay replied:
Ah, sorry, I thought that the order doesn't matter from the point of view of which result we arrive at, because it's a fixpoint loop. But this may not be true if `FoldConstants` and `NormalizeLets` fight with each other for some reason. I don't see a reason why they would fight, but anyhow I'll change the order back, because we are more robust that way, because normalization is important.

(The reason why I changed the order is that in the INSERT constant scenario (and probably in other similar cases) this would settle down with one less `NormalizeLets` run with the new order. But this is not so important.)

frankmcsherry reviewed (approved):
Looks good, still some comments, but I think this is better going forward. We can further optimize / specialize if we need, but I approve of starting from here, vs a more deeply specialized implementation!

ggevay commented:
>> So this fixes the performance for smaller inserts (1000 rows) but not with many rows (10000 rows), interesting.

> @antiguru observed that with large constants, much of the time ~of `FoldConstants`~ is in trying to determine if there is a column that forms a unique key.

Actually, this had a different reason, [see above](https://github.com/MaterializeInc/materialize/pull/30666#issuecomment-2511501579). (Btw., I think it worked for 10000 and didn't work for 100000, for the reason explained above.)

ggevay commented:
Nightly is fine:
- The relevant benchmark is showing a 20-30% speedup (compared to v0.125.3).
- The "Output consistency" failure is known (`ColumnKnowledge performed work after EquivalencePropagation`).
- The 0dt failure was due to a too small timeout in a test (https://github.com/MaterializeInc/materialize/pull/30693).
